akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -414,13 +456,13 @@ private[stream] object ConnectionSourceStage {
       new InHandler {
         override def onPush(): Unit = {
           val elem = grab(bytesIn)
+          elemCount += 1
           ReactiveStreamsCompliance.requireNonNullElement(elem)
           if (writeInProgress) {
             writeBuffer = writeBuffer ++ elem
           } else {
-            connection ! Write(writeBuffer ++ elem, WriteAck)
-            writeInProgress = true
-            writeBuffer = ByteString.empty
+            writeBuffer = writeBuffer ++ elem
```

Comment:
In which way is this case different than the `WriteAck` case? Wouldn't it mean that when a burst begins we always send an extra element that could also be coalesced?

Reply:
My thinking was that for low throughput scenarios we shouldn't delay anything. However, I think you said that for gRPC there are always more than 1 element (frame) so it could be worth trying the delay here also.